### ISSUE TITLE
sdk: sync Stream throws NPE when adding an event on onListen

### ISSUE THREAD

**Issue by frankpepermans** [dart-lang/sdk#41969]
We've encountered an issue with RxDart when trying to handle `startWith` for **sync** Streams:

In a nutshell, we simply just add the very first event the moment `onListen` happens, but when we do this, then a NPE error is thrown.

Internally, the `onListen` handler in stream.dart wants to pause the subscription, which is null at this point because it is lazily initialized (https://github.com/dart-lang/sdk/blob/master/sdk/lib/async/stream.dart#L520)

We could use `scheduleMicrotask` instead to add the first event, but then we encounter other issues and it is then not guaranteed that the `startWith` event is added before any other events, as this is after all a sync `Stream`.

Would a fix that simply does `subscription?.pause()` be enough?

**Comment by srawlins** [comment]
Thanks for filing an issue; can you provide a minimal reproduction case, and the error you see?

**Comment by frankpepermans** [comment]
@srawlins sure thing, you'll need to add rxdart 0.24.1 as a dependency:

first:
```
import 'dart:async';

import 'package:rxdart/rxdart.dart';


void main() async {
  final base = Stream.fromIterable(['']);
  final connectable = base.publishValue();
  connectable.connect();

  {
    final subscription = connectable.listen(null);
    await subscription.cancel();
  }
  {
    final subscription =
    connectable.asyncMap((event) async => event).listen(null);
    await subscription.cancel();
  }
}
```

second:
```
void main() async {
    await BehaviorSubject<int>.seeded(2, sync: true).first;
}
```

**Comment by frankpepermans** [comment]
The second example can also be triggered without rxdart:

```
void main() async {
  StreamController<int> controller;

  controller = StreamController<int>.broadcast(
      sync: true,
      onListen: () {
        controller.add(1);
      });

  await controller.stream.first;
}
```

**Comment by lrhn** [comment]
There is some bug here.

Adding a value in the `onListen` in synchronous broadcast stream will *not* deliver that value to the listener. The subscription has not been returned yet, so the listener is not considered active.
It might deliver to other existing listeners, though.
In this case, there are not existing listeners, so the event is lost.

For non-broadcast streams, early events are buffered, they are not for broadcast streams.
For asynchronous streams, the delivery is delayed until a later microtask, for synchronous streams they are not.
So, a synchronous broadcast controller will throw away events added before a listener becomes active, which is only after `onListen` has retuend.

It shouldn't throw! It also won't do what the code seems to expect.

So, I can fix this, but it won't make the event get delivered. Never add events in the `onListen` callback (or before, for that matter).

**Comment by lrhn** [comment]
@hoc081098 Yes, that is expected behavior.
Adding an event to a sync broadcast stream delivers it to all *existing* subscribers.

The one triggering the `onListen` is *not yet listening*. The `listen` call hasn't returned yet, the caller doesn't have access to the `StreamSubscription`. (I wish we hadn't provided callback parameters to `listen` and required you to set them on the returned subscription, because then it would be obvious that no-one is listening until `listen` has *returned*).

However, if this change to `first` made a difference, there might also be a bug in the broadcast stream code, because even before, it should not have worked.

For a non-broadcast stream, the event is buffered until the single possible listener is ready for them. 
Broadcast streams do not buffer events, they deliver them immediately to whoever is already listening. There is never a listener during the `onListen` call (unless the `onListen` call itself adds a listener, which I really can't recommend).

**Comment by hoc081098** [comment]
@lrhn Thanks for clear explanation.

**Comment by lrhn** [comment]
The `asyncMap` function has been rewritten since this, and doesn't have the problem any more.
